src/bio.h

```diff
@@ -30,13 +30,17 @@
 #ifndef __BIO_H
 #define __BIO_H
 
+typedef void lazy_free_fn(void *args[]);
+
 /* Exported API */
 void bioInit(void);
-void bioCreateBackgroundJob(int type, void *arg1, void *arg2, void *arg3);
 unsigned long long bioPendingJobsOfType(int type);
 unsigned long long bioWaitStepOfType(int type);
 time_t bioOlderJobOfType(int type);
 void bioKillThreads(void);
+void bioCreateCloseJob(int fd);
```

Comment:
Not a fan of this implementation.  The original BIO implementation is a mess, and this doesn't improve it much.

Conceptually, BIO shouldn't care about the jobs.  It should simply service a queue of jobs using a callback.

I recommend something like this:

```C
typedef void (*bioCallbackFunc)(void *data);

void bioCreateBackgroundJob(bioCallbackFunc callback, void *data);
```

This function would (internally) create a struct to hold the function pointer and callback data.  It would enqueue the item to be serviced by a background thread.  The background thread would simply call the callback function, and pass the (opaque) data.

The current BIO code is bad in that it's has to know about (and call) functions in the caller's space.  Rather than being a pure utility, a circular dependency is created.  The caller has to know about BIO - and BIO has to know about the caller.

Enhancements:
* The current code sets the thread title.  If that's desired, an additional parameter should be passed by the caller.
* The current code has the capability to return a count based on "type".  Instead, this should be based on the callback pointer. `long bioPendingJobsForCallback(bioCallbackFunc callback)`.
* It is possible, if desired, to internally break the requests into multiple queues so that different callbacks could be processed in parallel.  Example, having one queue that processes the CLOSE callback and another that processes the LAZYFREE callback.  (Of course BIO would only know that they are different callbacks, and have no idea what they actually do.)

Reply:
I disagree that this adds no value, your proposal doesn't add any other benefit besides being more general. The refactoring here does 2 main things:
1. More readable since now we don't have weird arguments in BIO jobs. 
2. It removes the need to extern lazy-free jobs, since those aren't generic.

It would still be net easier to decouple this one more step and move the lazyfree code into lazy free, and the aof/fsync code to their own files.